### PR TITLE
Allow trailing space (hyphen) in permalink while typing

### LINKED ISSUE #3455: Permalink input field doesn't allow hyphens anywhere, nor spaces at the end
## Bug Description

In the editor when manually changing the permalink, the entry field ignores hyphens that I type and spaces at the end of the entry string. So I can't enter something like "girl-hugs-lion" in the way it is written. I can insert hyphens by first entering "girlhugslion" and then go back and insert a space wherever I want the hyphen, since it translates the space to a hyphen. But what I also cannot do is directly type "girl hugs lion" and having it translate to "girl-hugs-lion", since spaces at the end of the field (and any char is the last char while you're first type the string) are ignored too.

## Expected Behaviour

In all other entry fields in WordPress all allowed permalink characters can also be entered directly in the entry-field and even none-allowed chars. Non allowed chars are filtered out,  and spaces translated to hyphens, only when the field is exited. So I should be able to just start typing in the permalink field using hyphens and spaces in the middle or at the end of the field and the system will translate when needed. When the plugin choses to validate during typing at least the allowed chars should be permitted.

## Steps to Reproduce

* Open the stories editor
* Go to the document option tab in the editor sidebar
* Enter a string containing spaces and/or hyphens in the permalink field.
* Notice that both hyphens and spaces are ignored while entering the string
* Try to insert a hyphen anywhere in the permalink
* Notice that no hyphen is inserted
* Try to insert a space anywhere in the permalink
* Notice that it is translated to a hyphen anywhere but at the end

Additional ways to reproduce issue:

1) Start a new story _without a title_.  One cannot add hyphens/ spaces within text until after the story is saved. [Video](https://recordit.co/dq74419Wtv)) 

2) Start a new story with  _a title_. One can add  hyphens/spaces within the pre-populated text.  ([Video](https://recordit.co/4yoinQqaUw))

In all cases,  adding hyphens/ spaces  _after_ the text is not possible. 

Possibly related detail: 

3) When using keys like space or underscore in the permalink entry field within the permalink value (so not at the end), the keystroke is translated to a hyphen, but the cursor is also moved to the end of the entry field. Expectation is that would remain on its position.

## Additional Context

- WordPress version: 5.4.2
- PHP version: 7.3.5
- Plugin Version:  1.0.0-beta.1
- Operating System: Windows 10
- Browser: Firefox Developer 79.0b9 (64-bits)

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

1. Go to document panel
1. Type something with spaces and/or dashes and observe, that it is possible (though both appear as dashes)
1. Make sure to leave a trailing hyphen, then blur the input field and observe, that the trailing hyphen is removed

### PR DESCRIPTION
**Pull request by barklund**
## Summary

This allows trailing (and leading) spaces (converted to hyphens) in the permalink input field while typing, but removes them on blur:

slug-hyphenated

## Relevant Technical Choices

* Add optional `allowTrailing` parameter to `cleanForSlug`
* Added blur handler to text input (and it needed access to the event too)

## To-do

- [x] Add optional `allowTrailing` parameter to `cleanForSlug`
- [x] Pass event object to `onBlur` handler in `TextInput`
- [x] Add blur handler in slug component
- [x] Add unit test for slug panel
- [x] Add unit test for`cleanForSlug` util function

## User-facing changes

* Creator can now type spaces and dashes at end of permalink while typing.

## Testing Instructions

* See if typing spaces and dashes at end of permalink while typing is possible.

---

Fixes #3455

### CODE REVIEW

**Comment by merapi** [comment]
Probably not a big deal since both links work and nobody sane will put multiple spaces (tho might happen when pasting some garbage?):

Type something with multiple spaces
image
After blur, now share that URL with somebody
image
After saving
image

**Comment by miina** [comment]
>Probably not a big deal since both links work and nobody sane will put multiple spaces 

Could also be a typo. Indeed would be great if the `cleanForSlug` would clean the duplicate hyphens within the slug as well.

File: assets/src/edit-story/utils/test/cleanForSlug.js [hunk 1]
```diff
@@ -32,14 +32,24 @@ describe('cleanForSlug', () => {
...
-    const result = cleanForSlug('-Hello-World-');
-    expect(result).toStrictEqual('hello-world');
+  it('should not trim leading and ending whitespace and dashes if editing', () => {
+    const result = cleanForSlug('- Hello-World- ', true);
+    expect(result).toStrictEqual('--hello-world--');
+  });
+
+  it('should collapse multiple whitespace and dashes', () => {
+    const result = cleanForSlug('Hello   cruel- world');
+    expect(result).toStrictEqual('hello-cruel-world');
```
Comment (miina) [inline assets/src/edit-story/utils/test/cleanForSlug.js:47]: 😮 Are you OK?

**Review by miina — APPROVED** [review]
LGTM

**Pull request merged**